### ISSUE TITLE
pyarrow

### ISSUE THREAD

**Issue by clstaudt** [pyodide/pyodide#2933]
## 🐍 Package Request

- Package Name and Version: pyarrow / latest
- Package URL: https://github.com/apache/arrow
- Package Dependencies that needs to be resolved first: ?

**Comment by clstaudt** [comment]
I'd be happy to contribute but this does not seem like an entry-level migration.

**Comment by hoodmane** [comment]
To start, you should try `python -m pyodide_build mkpkg pyarrow` and then `PYODIDE_PACKAGES=pyarrow make -C packages`. If that fails you can post the errors you get here and we will try to help fix them.

**Comment by hoodmane** [comment]
I guess it does sound a bit complicated:
https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst

**Comment by hoodmane** [comment]
You might have to hope that @ryanking13 takes an interest.

**Comment by clstaudt** [comment]
pyarrow is one option for .parquet file support, the other being fastparquet. Neither are in pyodide so far. The file format is really common though.

**Comment by hoodmane** [comment]
Fastparquet looks to me a bit simpler because the dependencies are also Python packages.

**Comment by Mause** [comment]
DuckDB is a possibility as well, as it has parquet support, and is already compilable to WASM -> https://github.com/duckdb/duckdb-wasm

(Full disclosure, I work on it 😅)

**Comment by hoodmane** [comment]
Does duckb have python bindings? It isn't obvious to me from the repos.

**Comment by Mause** [comment]
> Does duckb have python bindings? It isn't obvious to me from the repos.

Yep! https://pypi.org/project/duckdb/

**Comment by IanCal** [comment]
Duckdb builds without other issues, although I've not tried building the extensions. https://github.com/iancal/pyodide

Pyarrow would solve a lot of other issues though, including loading remote parquet files, and play nicely with duckdb so hopefully both can be included.

**Comment by changhiskhan** [comment]
Just want to toss in my two cents : would absolutely LOVE to see BOTH pyarrow and duckdb supported. pyarrow makes it super easy to integrate with other python tools in the ecosystem. And also it would give js tools a fast way to work with large amounts of remote data via arrowjs

**Comment by rth** [comment]
To give a status update on this, emscripten-forge does have [arrow packaged](https://github.com/emscripten-forge/recipes/blob/main/recipes/recipes_emscripten/arrow/recipe.yaml) by @DerThorsten so it does work with wasm. For pyodide someone would just need to do the work.

After talking with @jorisvandenbossche at JupyterCon, I quickly tried to build it for Pyodide, heavily inspired by what was done in emscripten-forge: https://github.com/pyodide/pyodide/compare/main...rth:pyodide:pyarrow?expand=1 but it clearly needs more cmake hacks and patches that I'm interested in doing on a weekend currently ) 

At the same time, it looks like @joemarshall was working on this upstream https://github.com/apache/arrow/issues/23221#issuecomment-1507172916 and https://github.com/apache/arrow/pull/35471. It's also possible that pyarrow would become a mandatory dependency in pandas in which case this would become critical https://github.com/pandas-dev/pandas/pull/52711

So in conclusion, contributing fixes upstream to reduce the number of patches necessary does seem to be the way to go. If @joemarshall would like to contribute the recipes that you have, it would also be very welcome.

**Comment by joemarshall** [comment]
Yeah, I'm waiting on that pr in arrow to get reviewed and in. Once that's in the build recipe for pyodide is trivial.

Without that patch, anything non trivial in arrow, especially anything that uses current arrow.dataset relies on threading and just breaks in emscripten if you don't build with threads. 

The problem is that whilst arrow has 'use_threads' options for a lot of things in many cases this only affects the CPU processing, and not the I/o thread pool, and in some cases it actually means 'do everything on one thread', but it launches a new thread to do the 'everything'.

Without the patch to properly disable threading in arrow core, things only work with a load of flags in various calls like 'use_legacy_dataset', and loads of potential calls in pyarrow cause crashes or hangs or just don't work. Oh and the pandas integration is mostly broken. It's not worth using in that state.

**Comment by rth** [comment]
Thats great, thanks for working on this and contributing fixes to arrow @joemarshall ! 

Also cc @jasongrout who was interested in it.

**Comment by kylebarron** [comment]
It's great to see @joemarshall 's threading PR was merged in Arrow yesterday https://github.com/apache/arrow/pull/35672!

**Comment by nicornk** [comment]
@joemarshall thanks for all your work so far!
Did you start creating the build recipe for pyodide already or are there any blockers?

**Comment by joemarshall** [comment]
> @joemarshall thanks for all your work so far!
> Did you start creating the build recipe for pyodide already or are there any blockers?

There's a couple of cmake / build script changes that I need to push to arrow first to fix cross compiling on that side. Once that's in, the build recipe should be relatively straightforward. I've done the libarrow stuff so it happily builds on emscripten, just need to do the python setup.py fixes.

**Comment by conker84** [comment]
@joemarshall sorry to bother you. What's the status of this?

**Comment by joemarshall** [comment]
Status is that the hard bit is nearly merged (wasm support in arrow), and the python side of things I did, but is waiting on the c++stuff to be merged. It dropped off my radar a bit because the contract that was in part paying me to work on it ran out. I'll finish it off when I have a bit of time, but paid work is a priority right now, so I can't guarantee how quick it will be.

**Comment by bitsondatadev** [comment]
> Status is that the hard bit is nearly merged (wasm support in arrow), and the python side of things I did, but is waiting on the c++stuff to be merged. It dropped off my radar a bit because the contract that was in part paying me to work on it ran out. I'll finish it off when I have a bit of time, but paid work is a priority right now, so I can't guarantee how quick it will be.

@joemarshall what's your going rate? With all the interest here, could we work on getting a bounty going? I'd love this to get done and you to get paid. Could you set up a GitHub sponsor and I'll go around asking for money like a kid on Halloween.

Also I know a few folks at Voltron who would likely chip in.

**Comment by joemarshall** [comment]
> > Status is that the hard bit is nearly merged (wasm support in arrow), and the python side of things I did, but is waiting on the c++stuff to be merged. It dropped off my radar a bit because the contract that was in part paying me to work on it ran out. I'll finish it off when I have a bit of time, but paid work is a priority right now, so I can't guarantee how quick it will be.
> 
> @joemarshall what's your going rate? With all the interest here, could we work on getting a bounty going? I'd love this to get done and you to get paid. Could you set up a GitHub sponsor and I'll go around asking for money like a kid on Halloween.
> 
> Also I know a few folks at Voltron who would likely chip in. 

Hi there,

I had a bit more time to work on it this week, status is that I need to poke arrow people to merge the c++ side, then the python stuff is a couple of pull requests away, one which will require discussion because it fixes some stuff about how the python build currently works and one uncontroversial one.

Money definitely helps in terms of keeping it in my high priority task list - I set up a GitHub sponsors link for this, if anyone wants to give a one of donation towards this port I'd definitely appreciate it.

https://github.com/sponsors/joemarshall

**Comment by joemarshall** [comment]
Status update, official C++ support for emscripten is merged now! Next job is updating and fixing up the python support. 

https://github.com/apache/arrow/pull/37821

Thanks ever so much to the people who have supported this work financially, it really helps me make the time to get this stuff done.

**Comment by joemarshall** [comment]
Right, current status of my build of pyarrow - I've done a fair bit of work on it, and finally got to a point where on my computer i build it and can run the full python test suite. 

There is a bugfix in emscripten that needs to go in ( https://github.com/emscripten-core/emscripten/pull/21759 - I can temporarily patch it in pyodide though, so it shouldn't hold anything up), along with one more fix to arrow c++ ( https://github.com/apache/arrow/pull/41125 ). 

There are a few test fails that still need fixing, and a few oddities relating to dates and times that probably just need documenting - JavaScript handles time and date slightly differently to Linux, and emscripten hands some of that stuff off to JavaScript.

I've also got to work out how to nicely script python tests in the arrow build, because pytest-pyodide doesn't work nicely with such a complex module structure...

**Comment by bitsondatadev** [comment]
@joemarshall, does the https://github.com/pyodide/pyodide/pull/4698 patch temporarily fix the emscripten issue in pyodide?

Are the final blockers seem to be single-threaded joins and the Python changes? Thanks :)

A reminder to anyone looking at this, while we have gotten Joe a good amount of compensation for this work, the extra hangups have him now going back into the red. Any donations to keep him going would be deeply appreciated!

https://github.com/sponsors/joemarshall

**Comment by joemarshall** [comment]
Yep, that and the tzdata pr that got merged into pyodide today also.

What this means is that the next pyodide release will be able to run arrow wheels without crashing.

The test suite is running fine when run manually on browser now, but some tests are failing when scripted, which is the last piece of the puzzle for integrating a pyodide build into ci and testing for arrow.

**Comment by cpcloud** [comment]
If the goal is to get `pyarrow` installable from the Pyodide build, just a heads up that it seems like adding any more long-building packages to the Pyodide CI is non-starter.

I've got a PR up for DuckDB (https://github.com/pyodide/pyodide/pull/4684) that seems like it will require a lot of CI refactoring to get in.

**Comment by bitsondatadev** [comment]
Ah got it, @ryanking13 @hoodmane would ot make sense to build duckdb/pyarrow out of tree for the next release?

**Comment by Alex-Monahan** [comment]
Is the Pyodide CI a funding question or engineering one?

**Comment by bitsondatadev** [comment]
Not one of the pyodide maintainers, but it is likely both. There seems to be some changes they want to make to the CI that will help build times and if they are hitting limits, they likely could use funding.

**Comment by hoodmane** [comment]
We get these hilarious emails from CircleCI saying we've exhausted the monthly OSS limits on like day 5 of each month and we need to pay or they will shut us off. The "limit" seems to be 20,000 compute minutes and 2 GB of storage per month. In February we used 171,135 compute minutes and 1.6 terabytes of storage. Their quote for how much we would have to pay to get this much compute/storage is $3,345.00 / month. But we've been doing this for years and they haven't shut us off yet. And it's not likely we're going to come up with $3345/month worth of funding any time soon.

**Comment by ryanking13** [comment]
> would ot make sense to build duckdb/pyarrow out of tree for the next release?

For the next Pyodide release, I think we can build duckdb/pyarrow out-of-tree + make a meta.yaml file in-tree that contains the URL that points to the out-of-tree wheel. This would allow the Pyodide distribution to include these packages without having to build them every time. 

```yaml
# meta.yaml file for pyarrow
package:
  name: pyarrow
source:
  url: https://<pyarrow_out_of_tree_wheel_url>.whl
```

**Comment by cpcloud** [comment]
For anyone interested, here's a **100% ephemeral** but working JupyterLite console with PyArrow, DuckDB, and Ibis installed and working together: https://pr-9009-9d4b9e39d2c4329cb6b4e03d838a755fc6f00826--ibis-quarto.netlify.app/repl. This will probably hit the Ibis website today or tomorrow.

**Comment by hoodmane** [comment]
> So now AFAICT, this pull request should be ready to go with a minor amount of back and forth at this point.

Would love to see this go forward! We could potentially backport it into 0.26.x to resolve #4840...

**Comment by bitsondatadev** [comment]
> > So now AFAICT, this pull request should be ready to go with a minor amount of back and forth at this point.
> 
> 
> 
> Would love to see this go forward! We could potentially backport it into 0.26.x to resolve #4840...

@hoodmane any idea on when we would need apache/arrow#37822 completed so we don't hold you up?

@joemarshall IIRC this should be the last PR on the Arrow side for all emscripten stuff needed?

cc: @cpcloud @ianmcook

**Comment by hoodmane** [comment]
I'd like to cut 0.26.1 around June 21st if possible. If we don't get this into 0.26.1 we can always release 0.26.2 with just pyarrow.

**Comment by bitsondatadev** [comment]
> I'd like to cut 0.26.1 around June 21st if possible. If we don't get this into 0.26.1 we can always release 0.26.2 with just pyarrow.

Hey @hoodmane https://github.com/apache/arrow/pull/37822 just got merged!

Huge thanks to @joemarshall for all his work here!

Any thoughts on when this can get in now? Or is the next train likely gonna be the next minor release?

**Comment by hoodmane** [comment]
Well I'm going to make 0.26.2 any day now...

**Comment by bitsondatadev** [comment]
> Well I'm going to make 0.26.2 any day now...

So this is either very timely or just too late. I according to @ianmcook, the next big Arrow release is going out and should include this.

Given the perceived demand for this feature, would it make sense to coordinate 0.26.2 just after the Arrow release, or possible do an early release with 0.27.0 or 0.26.3 shortly after so nothing is blocked?

**Comment by hoodmane** [comment]
I don't think it would be a problem to wait for 0.26.2. I don't think we'd release 0.27.0 as a package update -- @ryanking13 is actively trying to unvendor the package set so soon the releases for packages can be separated from releases for the runtime and ffi.

**Comment by bitsondatadev** [comment]
> I don't think it would be a problem to wait for 0.26.2. I don't think we'd release 0.27.0 as a package update -- @ryanking13 is actively trying to unvendor the package set so soon the releases for packages can be separated from releases for the runtime and ffi.

Fantastic. Also, here's Ian making the mention about the next major Arrow release: https://github.com/apache/arrow/pull/37822#issuecomment-2209424099

@ianmcook, where's the best location for pyodide folks to track the release status?

**Comment by isaacbrodsky** [comment]
It looks like the Arrow release has gone out now: https://arrow.apache.org/release/17.0.0.html

@cpcloud or others, could you share the package meta.yaml or other steps you took to build pyarrow for Pyodide? I ran into some odd error about `ReadError: Unknown archive format` when I naively tried to add a `meta.yaml`.

**Comment by cpcloud** [comment]
First order of business is probably get the emscripten wheels uploaded to the artifact repo so that they can be pulled from a pyodide meta.yaml.

**Comment by ianmcook** [comment]
cc @raulcd

**Comment by raulcd** [comment]
I am not sure if I understand the issue but pyarrow 17.0.0 is not yet on conda, see: https://github.com/conda-forge/pyarrow-feedstock/pull/130

**Comment by cpcloud** [comment]
@raulcd This isn't related to conda. The question is: where do the emscripten wheels for pyarrow live? Are they published anywhere?

**Comment by raulcd** [comment]
The only pyarrow wheels that have been published as part of the 17.0.0 release are the ones under: https://pypi.org/project/pyarrow/

The PR that added support for pyodide is this one: https://github.com/apache/arrow/pull/37822

I can't see any change on how the wheels were built on that PR.

Was the original idea to build new wheels for pyarrow-pyodide? Sorry, I wasn't on this discussion and I was not aware.
Also, should those wheels maintenance be part of the Arrow repository or should those be an outside package? I am not sure what the expectation is and what other projects do with pyodide specific wheels.

**Comment by hoodmane** [comment]
It doesn't look like https://github.com/apache/arrow/pull/37822 builds or tests in CI, let alone uploading the wheel somewhere.

**Comment by joemarshall** [comment]
Hi,

Emscripten is tested on the nightly ci, but wheels don't get published from that I think. 

There's a minor bit of patching to pyarrow required for timezones to work which lives in pyodide, because pyarrow doesn't distribute timezone data. I've got most of the work done, just fiddling with the meta.yaml to make it build inside pyodide build (it's a relatively complex build process). Libarrow builds fine, but pyarrow is throwing cmake errors which I need to work out. I've got quite a lot on right now but I am getting it done.

**Comment by cpcloud** [comment]
It seems like building in the pyodide CI might be a blocker, things time out on relatively frequent basis, but maybe @ryanking13 and/or @hoodmane would know more about that!

**Comment by joemarshall** [comment]
I've got a PR

**Comment by PavloPolovyi** [comment]
Hi everyone! Version 0.27 of pyodide was released and I was wondering how to make pyarrow work. When I call micropip.install I get an error `Can't find a pure Python 3 wheel for 'pyarrow'`. This is my code sample:

```
importScripts("https://cdn.jsdelivr.net/pyodide/v0.27.0/full/pyodide.js");

let pyodide;

async function init() {
    pyodide = await loadPyodide();
    pyodide.setDebug(false);
    await pyodide.loadPackage("micropip");
    const micropip = pyodide.pyimport("micropip");
    await micropip.install("pyarrow");
}
```
Could someone help me with this please? Maybe I'm doing something wrong?

**Comment by hoodmane** [comment]
Looks like it got disabled in the numpy v2 upgrade #4925 and we have reenabling it as a TODO. When we fix it, we can release Pyodide 0.27.1 with pyarrow.

**Comment by agriyakhetarpal** [comment]
Oops, thanks for noticing this, @PavloPolovyi. Apologies for the error; let me fix it in a moment.

**Comment by PavloPolovyi** [comment]
Thank you @agriyakhetarpal. So your fixes will be included in 0.27.1, right?

**Comment by agriyakhetarpal** [comment]
Yes!
